code/espurna/ota.ino

```diff
@@ -55,107 +69,123 @@ void _otaFrom(const char * host, unsigned int port, const char * url) {
     _ota_port = port;
     _ota_size = 0;
 
-    if (_ota_client == NULL) {
-        _ota_client = new AsyncClient();
-    }
+    bool connected;
 
-    _ota_client->onDisconnect([](void *s, AsyncClient *c) {
+    DEBUG_MSG_P(PSTR("[OTA] Connecting to %s:%u\n"), host, port);
 
-        DEBUG_MSG_P(PSTR("\n"));
-
-        if (Update.end(true)){
-            DEBUG_MSG_P(PSTR("[OTA] Success: %u bytes\n"), _ota_size);
-            deferredReset(100, CUSTOM_RESET_OTA);
+    #if ASYNC_TCP_SSL_ENABLED
+        if (port == 443) {
+            WiFiClientSecure _ota_client = _ota_client_secure;
+            connected = _ota_client.connect(host, port);
+            if (connected) {
+                char fp[60] = {0};
+                if (sslFingerPrintChar(getSetting("otafp", OTA_GITHUB_FP).c_str(), fp)) {
+                    #ifdef USING_AXTLS
+                    if (!_ota_client.verify(fp, host)) {
+                    #else
+                    if (!_ota_client.setFingerprint(fp)) {
+                    #endif
+                        DEBUG_MSG_P(PSTR("[OTA] Error: fingerprint doesn't match\n"));
+                        connected = false;
+                    }
+                } else {
+                    DEBUG_MSG_P(PSTR("[OTA] Error: wrong fingerprint\n"));
+                    connected = false;
+                }
+            }
         } else {
-            #ifdef DEBUG_PORT
-                Update.printError(DEBUG_PORT);
-            #endif
-            eepromRotate(true);
+            connected = _ota_client.connect(host, port);
         }
+    #else
+        connected = _ota_client.connect(host, port);
+    #endif
 
-        DEBUG_MSG_P(PSTR("[OTA] Disconnected\n"));
-
-        _ota_client->free();
-        delete _ota_client;
-        _ota_client = NULL;
-        free(_ota_host);
-        _ota_host = NULL;
-        free(_ota_url);
-        _ota_url = NULL;
-
-    }, 0);
-
-    _ota_client->onTimeout([](void *s, AsyncClient *c, uint32_t time) {
-        _ota_client->close(true);
-    }, 0);
-
-    _ota_client->onData([](void * arg, AsyncClient * c, void * data, size_t len) {
-
-        char * p = (char *) data;
-
-        if (_ota_size == 0) {
+    if (!connected) {
+        DEBUG_MSG_P(PSTR("[OTA] Connection failed\n"));
+        return;
+    }
 
-            Update.runAsync(true);
-            if (!Update.begin((ESP.getFreeSketchSpace() - 0x1000) & 0xFFFFF000)) {
-                #ifdef DEBUG_PORT
-                    Update.printError(DEBUG_PORT);
-                #endif
+    // Disabling EEPROM rotation to prevent writing to EEPROM after the upgrade
+    eepromRotate(false);
+
+    DEBUG_MSG_P(PSTR("[OTA] Downloading %s\n"), _ota_url);
+    char buffer[strlen_P(OTA_REQUEST_TEMPLATE) + strlen(_ota_url) + strlen(_ota_host)];
+    snprintf_P(buffer, sizeof(buffer), OTA_REQUEST_TEMPLATE, _ota_url, _ota_host);
+
+    _ota_client.print(buffer);
+
+    // Skip the response headers
```

Comment:
Any reason not to use HTTP client class? We can also check response code here.
Updater will fail verifying the data though... So maybe not needed?

Reply:
I looked into the ``HTTPClient`` class, it has some nice extra checks and allows to verify response codes etc - but I think it will only increase the build size and in this case the ``Update`` class will verify the response anyway. But if you think it's better to use it I'll change the code.

Reply:
Then, maybe later. Keep current state of things, when parsing is done by us not by some HTTP lib.
But I'd double-check that updater does not needlessly write things to flash